rfcs/exact-table-types.md

```diff
@@ -0,0 +1,245 @@
+connotation, but for exactness. One proposal discussed in the alternatives would be to follow Flow
+in using `{| x: number, y: number |}` to mean a sealed, exact table.
+
+Though we are proposing to make the default interpretation of table _annotations_ sealed and exact,
+we maintain that the sensible default inference behavior is to default to inferring inexact table
+types for unannotated table parameters. Consider the following example:
+
+```lua
+function dostuff(t)
+  ...
+  t.x = 4
+  ...
+  t.y = 5
+  ...
+end
+```
+
+In this case, we expect that `t` will be given the inferred type `{x: number, y: number, ...}`.
+However, if we also passed `t` to another function that requires an exact table, we can propagate
+this requirement to `dostuff`.
+
+```lua
+function dostuff(t)
+  ...
+  t.x = 4
+  ...
+  t.y = 5
+  ...
+  dootherstuff(t)
+  ...
+end
+
+function dootherstuff(pt: {x: number, y: number})
+  ...
+end
+```
+
+So, in this case, `t` would be given the exact table type `{x: number, y: number}` during inference.
+Note that this is still true even if `dostuff` did not refer to some or all of the properties of `t`
+outside of passing it into `dootherstuff`. That is, the type will be the same even if the code was
+as follows:
+
+```lua
+function dostuff(t)
+  return dootherstuff(t)
+end
+
+function dootherstuff(pt: {x: number, y: number})
+  ...
+end
+```
+
+If instead the table is passed to a function requiring an exact table type with _different_
+properties than the ones used in `dostuff`, we will emit an error indicating that they are in
+conflict.
+
+```lua
+function dostuff(t)
+  ...
+  t.z = 4
+  ...
+  ...
+  dootherstuff(t)
+  ...
+end
+
+function dootherstuff(pt: {x: number, y: number})
+  ...
+end
+```
+
+In this case, `t` has two constraints: one indicating that it should be the exact table type
+`{x: number, y: number}` and the other indicating that it should have the property `z`. These
+constraints cannot be reconciled, and so we will get an error at the call site of `dootherstuff`.
+
+## Drawbacks
+
+The main drawback of doing this is that the existing RFC for [sealed table
+subtyping][width-subtyping] was already implemented as of March 1st, meaning that this is ultimately
+a breaking change. As such, it would be important to manage the change as a gradual rollout to
+mitigate problems. Since Luau currently only has sealed, inexact table types we can start by adding
+the syntax for explicit inexact tables (e.g. `{x: number, y: number, ...}`) and implement a warning
+that explains the default behavior will change in the future and to change it to the explicit
+inexact syntax if the programmer would like to keep the current behavior. At a later date, we can
+enable exact-by-default, but should provide an explicit suggestion in the error message to tell the
+programmer that if they _intended_ to allow tables with additional properties, they can change the
+signature of the function to include the `...`.
```

Comment:
Maybe the discussion about inexact tables should be more prominent than just a section inside "Drawbacks", or is this intended to go in a different RFC? It feels like an important point

e.g., Personally, I have no particular opinion, but were alternatives in syntax considered (such as TypeScript's `interface { }`)

Reply:
There is mention in the design section about the syntax of inexact tables. Is there something you're concerned isn't discussed? I can, at minimum, say explicitly that the behavior of inexact tables defined using the proposed inexact syntax would be the same as the current behavior (i.e. they support the width subtyping already accepted and implemented in #38.)

Reply:
Ah yeah, I see it now - I must have missed it earlier.

No particular issue for me, just a note that it is very likely to get comments down the road from users about "why was XYZ not chosen", especially for syntax.